hal/stm32c0.c

```diff
@@ -0,0 +1,323 @@
+}
+
+int RAMFUNCTION hal_flash_write(uint32_t address, const uint8_t *data, int len)
+{
+    int i = 0;
+    uint32_t *src, *dst;
+    flash_clear_errors();
+    FLASH_CR |= FLASH_CR_PG;
+
+    while (i < len) {
+        flash_clear_errors();
+        if ((len - i > 3) && ((((address + i) & 0x07) == 0) &&
+                ((((uint32_t)data) + i) & 0x07) == 0)) {
+            src = (uint32_t *)data;
+            dst = (uint32_t *)(address + FLASHMEM_ADDRESS_SPACE);
+            flash_wait_complete();
+            dst[i >> 2] = src[i >> 2];
+            dst[(i >> 2) + 1] = src[(i >> 2) + 1];
+            flash_wait_complete();
+            i+=8;
+        } else {
+            uint32_t val[2];
+            uint8_t *vbytes = (uint8_t *)(val);
+            int off = (address + i) - (((address + i) >> 3) << 3);
+            uint32_t base_addr = address & (~0x07); /* aligned to 64 bit */
+            int u32_idx = (i >> 2);
+            dst = (uint32_t *)(base_addr);
+            val[0] = dst[u32_idx];
+            val[1] = dst[u32_idx + 1];
+            while ((off < 8) && (i < len))
+                vbytes[off++] = data[i++];
+            dst[u32_idx] = val[0];
+            dst[u32_idx + 1] = val[1];
+            flash_wait_complete();
+        }
+    }
+    if ((FLASH_SR & FLASH_SR_EOP) == FLASH_SR_EOP)
+        FLASH_SR |= FLASH_SR_EOP;
+    FLASH_CR &= ~FLASH_CR_PG;
+    return 0;
+}
+
+void RAMFUNCTION hal_flash_unlock(void)
+{
+    flash_wait_complete();
+    if ((FLASH_CR & FLASH_CR_LOCK) != 0) {
+        FLASH_KEY = FLASH_KEY1;
+        DMB();
+        FLASH_KEY = FLASH_KEY2;
+        DMB();
+        while ((FLASH_CR & FLASH_CR_LOCK) != 0)
+            ;
+    }
+}
+
+void RAMFUNCTION hal_flash_lock(void)
+{
+    flash_wait_complete();
+    if ((FLASH_CR & FLASH_CR_LOCK) == 0)
+        FLASH_CR |= FLASH_CR_LOCK;
+}
+
+
+int RAMFUNCTION hal_flash_erase(uint32_t address, int len)
+{
+    int start = -1, end = -1;
+    uint32_t end_address;
+    uint32_t p;
+    if (len == 0)
+        return -1;
+    end_address = address + len - 1;
+    for (p = address; p < end_address; p += FLASH_PAGE_SIZE) {
+        uint32_t reg = FLASH_CR & (~(FLASH_CR_PNB_MASK << FLASH_CR_PNB_SHIFT));
+        FLASH_CR = reg | ((p >> FLASH_PAGE_SIZE_SHIFT) << FLASH_CR_PNB_SHIFT) | FLASH_CR_PER;
+        DMB();
+        FLASH_CR |= FLASH_CR_STRT;
+        flash_wait_complete();
+        FLASH_CR &= ~FLASH_CR_PER;
+    }
+    return 0;
+}
+
+static void clock_pll_off(void)
+{
+    uint32_t reg32;
+
+    /* Select HSISYS as SYSCLK source. */
+    reg32 = RCC_CFGR;
+    reg32 &= ~((1 << 1) | (1 << 0));
+    RCC_CFGR = (reg32 | RCC_CFGR_SW_HSISYS);
+    DMB();
+}
+
+/* This implementation will setup HSI RC 48 MHz as System Clock Source, set
+ * flash wait state to 1, and set all peripherals to 16MHz (div4)
+ */
+static void clock_pll_on(int powersave)
+{
+    uint32_t reg32;
+    uint32_t cpu_freq, plln, pllm, pllq, pllp, pllr, hpre, ppre, flash_waitstates;
```

Comment:
Remove declarations of unused variables (cpu_freq, plln, pllm, pllq, pllp, pllr, hpre, ppre)